src/mlpack/methods/dbscan/dbscan_impl.hpp

```diff
@@ -164,53 +184,62 @@ void DBSCAN<RangeSearchType, PointSelectionPolicy>::PointwiseCluster(
   // We maintain a list of non-core points so that we can handle that logic
   // correctly.
 
-  std::vector<bool> visited(data.n_cols, false);
-  std::vector<bool> nonCorePoints(data.n_cols, false);
+  std::vector<std::atomic<bool>> visited(data.n_cols);
+  std::vector<std::atomic<bool>> nonCorePoints(data.n_cols);
 
   for (size_t i = 0; i < data.n_cols; ++i)
   {
-    if (i % 10000 == 0 && i > 0)
-      Log::Info << "DBSCAN clustering on point " << i << "..." << std::endl;
-
-    // Get the next index.
-    const size_t index = pointSelector.Select(i, data);
-    visited[index] = true;
+    visited[i] = false;
+    nonCorePoints[i] = false;
+  }
 
-    // Do the range search for only this point.
-    rangeSearch.Search(data.col(index), RangeType<ElemType>(ElemType(0.0), epsilon),
-        neighbors, distances);
+  #pragma omp parallel
+  {
+    std::vector<std::vector<size_t>> neighbors;
+    std::vector<std::vector<ElemType>> distances;
 
-    // Union to all neighbors if the point is not noise.
-    //
-    // If the point is noise, we leave its label as undefined (i.e. we do no
-    // unioning).
-    if (neighbors[0].size() >= minPoints)
+    #pragma omp for schedule(dynamic)
+    for (size_t i = 0; i < data.n_cols; ++i)
     {
-      for (size_t j = 0; j < neighbors[0].size(); ++j)
+      if (i % 10000 == 0 && i > 0)
+        Log::Info << "DBSCAN clustering on point " << i << "..." << std::endl;
+
+      // Get the next index.
+      const size_t index = pointSelector.Select(i, data);
+      visited[index] = true;
+
+      // Do the range search for only this point.
+      rangeSearch.Search(data.col(index), RangeType<ElemType>(ElemType(0.0), epsilon),
```

Comment:
Realistically this is the "painful" part of DBSCAN. We would likely get better results by parallelizing the range search itself, but that is a bit of a can of worms, since you would have to parallelize the tree traversal (likely with OpenMP tasks) and then ensure that `RangeSearchRules` is thread-safe.  Not impossible for sure, but probably a good amount of work and tuning.